src/zcash/NoteEncryption.hpp

```diff
@@ -87,9 +92,34 @@ class note_decryption_failed : public std::runtime_error {
     note_decryption_failed() : std::runtime_error("Could not decrypt message") { }
 };
 
+
+
+// Subclass PaymentDisclosureNoteDecryption provides a method to decrypt a note with esk.
+template<size_t MLEN>
+class PaymentDisclosureNoteDecryption : public NoteDecryption<MLEN> {
+protected:
+public:
+    enum { CLEN=MLEN+NOTEENCRYPTION_AUTH_BYTES };
+    typedef boost::array<unsigned char, CLEN> Ciphertext;
+    typedef boost::array<unsigned char, MLEN> Plaintext;
+
+    PaymentDisclosureNoteDecryption() : NoteDecryption<MLEN>() {}
+    PaymentDisclosureNoteDecryption(uint256 sk_enc) : NoteDecryption<MLEN>(sk_enc) {}
+
+    Plaintext decryptWithEsk(
+        const Ciphertext &ciphertext,
+        const uint256 &pk_enc,
+        const uint256 &esk,
+        const uint256 &hSig,
+        unsigned char nonce
+        ) const;
```

Comment:
I would have been inclined to just add ``decryptWithEsk`` to ``NoteDecryption``, for simplicity. It doesn't really matter though.

Reply:
That's how I did it originally... but then I changed my mind as I didn't want to pollute the existing class.